pkg/odo/cli/init/init.go

```diff
@@ -0,0 +1,319 @@
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Set the name in the devfile *AND* writes the devfile back to the disk in case
+	// it has been removed and not replaced by the starter project
+	err = devfileObj.SetMetadataName(o.InitParams.Name)
+	if err != nil {
+		return fmt.Errorf("Failed to update the devfile's name: %w", err)
+	}
+
+	log.Italicf(`
+Your new component %q is ready in the current directory.
+To start editing your component, use "odo dev" and open this folder in your favorite IDE.
+Changes will be directly reflected on the cluster.
+To deploy your component to a cluster use "odo deploy".`, o.InitParams.Name)
+
+	return nil
+}
+
+// downloadDirect downloads a devfile at the provided URL and saves it in dest
+func (o *InitOptions) downloadDirect(URL string, dest string) error {
+	parsedURL, err := url.Parse(URL)
+	if err != nil {
+		return err
+	}
+	if strings.HasPrefix(parsedURL.Scheme, "http") {
+		downloadSpinner := log.Spinnerf("Downloading devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		params := util.HTTPRequestParams{
+			URL: URL,
+		}
+		devfileData, err := o.registryClient.DownloadFileInMemory(params)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, devfileData, 0644)
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	} else {
+		downloadSpinner := log.Spinnerf("Copying devfile from %q", URL)
+		defer downloadSpinner.End(false)
+		content, err := o.fsys.ReadFile(URL)
+		if err != nil {
+			return err
+		}
+		info, err := o.fsys.Stat(URL)
+		if err != nil {
+			return err
+		}
+		err = o.fsys.WriteFile(dest, content, info.Mode().Perm())
+		if err != nil {
+			return err
+		}
+		downloadSpinner.End(true)
+	}
+
+	return nil
+}
+
+// downloadFromRegistry downloads a devfile from the provided registry and saves it in dest
+// If registryName is empty, will try to download the devfile from the list of registries in preferences
+func (o *InitOptions) downloadFromRegistry(registryName string, devfile string, dest string) error {
+	var downloadSpinner *log.Status
+	var forceRegistry bool
+	if registryName == "" {
+		downloadSpinner = log.Spinnerf("Downloading devfile %q", devfile)
+		forceRegistry = false
+	} else {
+		downloadSpinner = log.Spinnerf("Downloading devfile %q from registry %q", devfile, registryName)
+		forceRegistry = true
+	}
+	defer downloadSpinner.End(false)
+
+	registries := o.preferenceClient.RegistryList()
+	var reg preference.Registry
+	for _, reg = range *registries {
+		if forceRegistry && reg.Name == registryName {
+			err := o.registryClient.PullStackFromRegistry(reg.URL, devfile, dest, segment.GetRegistryOptions())
+			if err != nil {
+				return err
+			}
+			downloadSpinner.End(true)
+			return nil
+		} else if !forceRegistry {
+			err := o.registryClient.PullStackFromRegistry(reg.URL, devfile, dest, segment.GetRegistryOptions())
+			if err != nil {
+				continue
+			}
+			downloadSpinner.End(true)
+			return nil
+		}
+	}
```

Comment:
Can you explain this bit of code? If registry is not forced, then we pull from every registry available?

Reply:
When the user does not provides a specific registry, the devfile is pulled from the first registry containing this devfile (see https://github.com/redhat-developer/odo/issues/5297#issuecomment-1018463900)